servant/src/Servant/API/Verbs.hs

```diff
@@ -29,12 +29,17 @@ import           Network.HTTP.Types.Method
 data Verb (method :: k1) (statusCode :: Nat) (contentTypes :: [*]) (a :: *)
   deriving (Typeable, Generic)
 
--- | @NoContentVerb@ is a specific type to represent 'NoContent' responses.
--- It does not require either a list of content types (because there's
--- no content) or a status code (because it should always be 204).
-data NoContentVerb  (method :: k1)
+-- | @NoContentVerbWithStatus@ is a specific type to represent 'NoContent' responses.
+-- It does not require either a list of content types (because there's no content).
+-- It still requires a status code, because several statuses may have no content.
+-- (e.g. 204, 301, 302, or 303).
```

Comment:
I do have a philosophical question here: when using 30X status codes, the `Location` header must be set. We are not enforcing this here — and in fact, your test does not do that either.

Wouldn't it be better to provide redirect combinators that do enforce this constraint ?